File: trunk/java/edu/nps/moves/dis/DetonationPdu.java
package edu.nps.moves.dis;

import java.util.*;
import java.io.*;

/**
 * Section 5.3.4.2. Information about stuff exploding. COMPLETE
 *
 * Copyright (c) 2007, MOVES Institute, Naval Postgraduate School. All rights reserved.
 * This work is licensed under the BSD open source license, available at https://www.movesinstitute.org/licenses/bsd.html
 *
 * @author DMcG
 */
public class DetonationPdu extends Warfare
{
   /** ID of muntion that was fired */
   protected EntityID  munitionID = new EntityID(); 

   /** ID firing event */
   protected EventID  eventID = new EventID(); 

   /** ID firing event */
   protected Vector3Float  velocity = new Vector3Float(); 

   /** where the detonation is, in world coordinates */
   protected Vector3Double  locationInWorldCoordinates = new Vector3Double(); 

   /** Describes munition used */
   protected BurstDescriptor  burstDescriptor = new BurstDescriptor(); 

   /** result of the explosion */
   protected short  detonationResult;

   /** How many articulation parameters we have */
   protected short  numberOfArticulationParameters;

   /** padding */
   protected short  pad = 0;

   protected List articulationParameters = new ArrayList(); 

/** Constructor */
 public DetonationPdu()
 {
    setPduType( (short)3 );
 }

public int getMarshalledSize()
{
   int marshalSize = 0; 

   marshalSize = super.getMarshalledSize();
   marshalSize = marshalSize + munitionID.getMarshalledSize();  // munitionID
   marshalSize = marshalSize + eventID.getMarshalledSize();  // eventID
   marshalSize = marshalSize + velocity.getMarshalledSize();  // velocity
   marshalSize = marshalSize + locationInWorldCoordinates.getMarshalledSize();  // locationInWorldCoordinates
   marshalSize = marshalSize + burstDescriptor.getMarshalledSize();  // burstDescriptor
   marshalSize = marshalSize + 1;  // detonationResult
   marshalSize = marshalSize + 1;  // numberOfArticulationParameters
   marshalSize = marshalSize + 2;  // pad
   for(int idx=0; idx < articulationParameters.size(); idx++)
   {
        ArticulationParameter listElement = (ArticulationParameter)articulationParameters.get(idx);
        marshalSize = marshalSize + listElement.getMarshalledSize();
   }

   return marshalSize;
}


public void setMunitionID(EntityID pMunitionID)
{ munitionID = pMunitionID;
}

public EntityID getMunitionID()
{ return munitionID; }

public void setEventID(EventID pEventID)
{ eventID = pEventID;
}

public EventID getEventID()
{ return eventID; }

public void setVelocity(Vector3Float pVelocity)
{ velocity = pVelocity;
}

public Vector3Float getVelocity()
{ return velocity; }

public void setLocationInWorldCoordinates(Vector3Double pLocationInWorldCoordinates)
{ locationInWorldCoordinates = pLocationInWorldCoordinates;
}

public Vector3Double getLocationInWorldCoordinates()
{ return locationInWorldCoordinates; }

public void setBurstDescriptor(BurstDescriptor pBurstDescriptor)
{ burstDescriptor = pBurstDescriptor;
}

public BurstDescriptor getBurstDescriptor()
{ return burstDescriptor; }

public void setDetonationResult(short pDetonationResult)
{ detonationResult = pDetonationResult;
}

public short getDetonationResult()
{ return detonationResult; 
}

public short getNumberOfArticulationParameters()
{ return (short)articulationParameters.size();
}

public void setPad(short pPad)
{ pad = pPad;
}

public short getPad()
{ return pad; 
}

public void setArticulationParameters(List pArticulationParameters)
{ articulationParameters = pArticulationParameters;
}

public List getArticulationParameters()
{ return articulationParameters; }


public void marshal(DataOutputStream dos)
{
    super.marshal(dos);
    try 
    {
       munitionID.marshal(dos);
       eventID.marshal(dos);
       velocity.marshal(dos);
       locationInWorldCoordinates.marshal(dos);
       burstDescriptor.marshal(dos);
       dos.writeByte( (byte)detonationResult);
       dos.writeByte( (byte)articulationParameters.size());
       dos.writeShort( (short)pad);

       for(int idx = 0; idx < articulationParameters.size(); idx++)
       {
            ArticulationParameter aArticulationParameter = (ArticulationParameter)articulationParameters.get(idx);
            aArticulationParameter.marshal(dos);
       } // end of list marshalling

    } // end try 
    catch(Exception e)
    { 
      System.out.println(e);}
    } // end of marshal method

public void unmarshal(DataInputStream dis)
{
    super.unmarshal(dis);

    try 
    {
       munitionID.unmarshal(dis);
       eventID.unmarshal(dis);
       velocity.unmarshal(dis);
       locationInWorldCoordinates.unmarshal(dis);
       burstDescriptor.unmarshal(dis);
       detonationResult = dis.readByte();
       numberOfArticulationParameters = dis.readByte();
       pad = dis.readShort();
        for(int idx = 0; idx < numberOfArticulationParameters; idx++)
        {
           ArticulationParameter anX = new ArticulationParameter();
            anX.unmarshal(dis);
            articulationParameters.add(anX);
        };

    } // end try 
   catch(Exception e)
    { 
      System.out.println(e); 
    }
 } // end of unmarshal method 


 /**
  * The equals method doesn't always work--mostly on on classes that consist only of primitives. Be careful.
  */
 public boolean equals(DetonationPdu rhs)
 {
     boolean ivarsEqual = true;

    if(rhs.getClass() != this.getClass())
        return false;

     if( ! (munitionID.equals( rhs.munitionID) )) ivarsEqual = false;
     if( ! (eventID.equals( rhs.eventID) )) ivarsEqual = false;
     if( ! (velocity.equals( rhs.velocity) )) ivarsEqual = false;
     if( ! (locationInWorldCoordinates.equals( rhs.locationInWorldCoordinates) )) ivarsEqual = false;
     if( ! (burstDescriptor.equals( rhs.burstDescriptor) )) ivarsEqual = false;
     if( ! (detonationResult == rhs.detonationResult)) ivarsEqual = false;
     if( ! (numberOfArticulationParameters == rhs.numberOfArticulationParameters)) ivarsEqual = false;
     if( ! (pad == rhs.pad)) ivarsEqual = false;

     for(int idx = 0; idx < articulationParameters.size(); idx++)
     {
        ArticulationParameter x = (ArticulationParameter)articulationParameters.get(idx);
        if( ! ( articulationParameters.get(idx).equals(rhs.articulationParameters.get(idx)))) ivarsEqual = false;
     }


    return ivarsEqual;
 }
} // end of class

File: trunk/open-dis/Csharp/DIS/DetonationPdu.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Diagnostics;
using System.Xml.Serialization;
using DISnet.DataStreamUtilities;

namespace DIS1998net
{

/**
 * Section 5.3.4.2. Information about stuff exploding. COMPLETE
 *
 * Copyright (c) 2008, MOVES Institute, Naval Postgraduate School. All rights reserved.
 * This work is licensed under the BSD open source license, available at https://www.movesinstitute.org/licenses/bsd.html
 *
 * @author DMcG
 * Modified for use with C#:
 * Peter Smith (Naval Air Warfare Center - Training Systems Division)
 */
[Serializable]
[XmlRoot]
[XmlInclude(typeof(EntityID))]
[XmlInclude(typeof(EventID))]
[XmlInclude(typeof(Vector3Float))]
[XmlInclude(typeof(Vector3Double))]
[XmlInclude(typeof(BurstDescriptor))]
[XmlInclude(typeof(ArticulationParameter))]
public class DetonationPdu : WarfareFamilyPdu
{
   /** ID of muntion that was fired */
   protected EntityID  _munitionID = new EntityID(); 

   /** ID firing event */
   protected EventID  _eventID = new EventID(); 

   /** ID firing event */
   protected Vector3Float  _velocity = new Vector3Float(); 

   /** where the detonation is, in world coordinates */
   protected Vector3Double  _locationInWorldCoordinates = new Vector3Double(); 

   /** Describes munition used */
   protected BurstDescriptor  _burstDescriptor = new BurstDescriptor(); 

   /** location of the detonation or impact in the target entity's coordinate system. This information should be used for damage assessment. */
   protected Vector3Float  _locationInEntityCoordinates = new Vector3Float(); 

   /** result of the explosion */
   protected byte  _detonationResult;

   /** How many articulation parameters we have */
   protected byte  _numberOfArticulationParameters;

   /** padding */
   protected short  _pad = 0;

   protected List<object> _articulationParameters = new List<object>(); 

/** Constructor */
 public DetonationPdu()
 {
    PduType = (byte)3;
 }

public int getMarshalledSize()
{
   int marshalSize = 0; 

   marshalSize = base.getMarshalledSize();
   marshalSize = marshalSize + _munitionID.getMarshalledSize();  // _munitionID
   marshalSize = marshalSize + _eventID.getMarshalledSize();  // _eventID
   marshalSize = marshalSize + _velocity.getMarshalledSize();  // _velocity
   marshalSize = marshalSize + _locationInWorldCoordinates.getMarshalledSize();  // _locationInWorldCoordinates
   marshalSize = marshalSize + _burstDescriptor.getMarshalledSize();  // _burstDescriptor
   marshalSize = marshalSize + _locationInEntityCoordinates.getMarshalledSize();  // _locationInEntityCoordinates
   marshalSize = marshalSize + 1;  // _detonationResult
   marshalSize = marshalSize + 1;  // _numberOfArticulationParameters
   marshalSize = marshalSize + 2;  // _pad
   for(int idx=0; idx < _articulationParameters.Count; idx++)
   {
        ArticulationParameter listElement = (ArticulationParameter)_articulationParameters[idx];
        marshalSize = marshalSize + listElement.getMarshalledSize();
   }

   return marshalSize;
}


public void setMunitionID(EntityID pMunitionID)
{ _munitionID = pMunitionID;
}

public EntityID getMunitionID()
{ return _munitionID; 
}

[XmlElement(Type= typeof(EntityID), ElementName="munitionID")]
public EntityID MunitionID
{
     get
{
          return _munitionID;
}
     set
{
          _munitionID = value;
}
}

public void setEventID(EventID pEventID)
{ _eventID = pEventID;
}

public EventID getEventID()
{ return _eventID; 
}

[XmlElement(Type= typeof(EventID), ElementName="eventID")]
public EventID EventID
{
     get
{
          return _eventID;
}
     set
{
          _eventID = value;
}
}

public void setVelocity(Vector3Float pVelocity)
{ _velocity = pVelocity;
}

public Vector3Float getVelocity()
{ return _velocity; 
}

[XmlElement(Type= typeof(Vector3Float), ElementName="velocity")]
public Vector3Float Velocity
{
     get
{
          return _velocity;
}
     set
{
          _velocity = value;
}
}

public void setLocationInWorldCoordinates(Vector3Double pLocationInWorldCoordinates)
{ _locationInWorldCoordinates = pLocationInWorldCoordinates;
}

public Vector3Double getLocationInWorldCoordinates()
{ return _locationInWorldCoordinates; 
}

[XmlElement(Type= typeof(Vector3Double), ElementName="locationInWorldCoordinates")]
public Vector3Double LocationInWorldCoordinates
{
     get
{
          return _locationInWorldCoordinates;
}
     set
{
          _locationInWorldCoordinates = value;
}
}

public void setBurstDescriptor(BurstDescriptor pBurstDescriptor)
{ _burstDescriptor = pBurstDescriptor;
}

public BurstDescriptor getBurstDescriptor()
{ return _burstDescriptor; 
}

[XmlElement(Type= typeof(BurstDescriptor), ElementName="burstDescriptor")]
public BurstDescriptor BurstDescriptor
{
     get
{
          return _burstDescriptor;
}
     set
{
          _burstDescriptor = value;
}
}

public void setLocationInEntityCoordinates(Vector3Float pLocationInEntityCoordinates)
{ _locationInEntityCoordinates = pLocationInEntityCoordinates;
}

public Vector3Float getLocationInEntityCoordinates()
{ return _locationInEntityCoordinates; 
}

[XmlElement(Type= typeof(Vector3Float), ElementName="locationInEntityCoordinates")]
public Vector3Float LocationInEntityCoordinates
{
     get
{
          return _locationInEntityCoordinates;
}
     set
{
          _locationInEntityCoordinates = value;
}
}

public void setDetonationResult(byte pDetonationResult)
{ _detonationResult = pDetonationResult;
}

[XmlElement(Type= typeof(byte), ElementName="detonationResult")]
public byte DetonationResult
{
     get
{
          return _detonationResult;
}
     set
{
          _detonationResult = value;
}
}

public void setPad(short pPad)
{ _pad = pPad;
}

[XmlElement(Type= typeof(short), ElementName="pad")]
public short Pad
{
     get
{
          return _pad;
}
     set
{
          _pad = value;
}
}

public void setArticulationParameters(List<object> pArticulationParameters)
{ _articulationParameters = pArticulationParameters;
}

public List<object> getArticulationParameters()
{ return _articulationParameters; }

[XmlElement(ElementName = "articulationParametersList",Type = typeof(List<object>))]
public List<object> ArticulationParameters
{
     get
{
          return _articulationParameters;
}
     set
{
          _articulationParameters = value;
}
}


public void marshal(DataOutputStream dos)
{
    base.marshal(dos);
    try 
    {
       _munitionID.marshal(dos);
       _eventID.marshal(dos);
       _velocity.marshal(dos);
       _locationInWorldCoordinates.marshal(dos);
       _burstDescriptor.marshal(dos);
       _locationInEntityCoordinates.marshal(dos);
       dos.writeByte( (byte)_detonationResult);
       dos.writeByte( (byte)_articulationParameters.Count);
       dos.writeShort( (short)_pad);

       for(int idx = 0; idx < _articulationParameters.Count; idx++)
       {
            ArticulationParameter aArticulationParameter = (ArticulationParameter)_articulationParameters[idx];
            aArticulationParameter.marshal(dos);
       } // end of list marshalling

    } // end try 
    catch(Exception e)
    { 
      Trace.WriteLine(e);
      Trace.Flush();
    }
} // end of marshal method

public void unmarshal(DataInputStream dis)
{
    base.unmarshal(dis);

    try 
    {
       _munitionID.unmarshal(dis);
       _eventID.unmarshal(dis);
       _velocity.unmarshal(dis);
       _locationInWorldCoordinates.unmarshal(dis);
       _burstDescriptor.unmarshal(dis);
       _locationInEntityCoordinates.unmarshal(dis);
       _detonationResult = dis.readByte();
       _numberOfArticulationParameters = dis.readByte();
       _pad = dis.readShort();
        for(int idx = 0; idx < _numberOfArticulationParameters; idx++)
        {
           ArticulationParameter anX = new ArticulationParameter();
            anX.unmarshal(dis);
            _articulationParameters.Add(anX);
        };

    } // end try 
   catch(Exception e)
    { 
      Trace.WriteLine(e); 
      Trace.Flush();
    }
 } // end of unmarshal method 


public void reflection(StringBuilder sb)
{
    sb.Append("----- DetonationPdu-----"  + System.Environment.NewLine);
    base.reflection(sb);
    try 
    {
       sb.Append("=====_munitionID=====" + System.Environment.NewLine);
       _munitionID.reflection(sb);
       sb.Append("=====_eventID=====" + System.Environment.NewLine);
       _eventID.reflection(sb);
       sb.Append("=====_velocity=====" + System.Environment.NewLine);
       _velocity.reflection(sb);
       sb.Append("=====_locationInWorldCoordinates=====" + System.Environment.NewLine);
       _locationInWorldCoordinates.reflection(sb);
       sb.Append("=====_burstDescriptor=====" + System.Environment.NewLine);
       _burstDescriptor.reflection(sb);
       sb.Append("=====_locationInEntityCoordinates=====" + System.Environment.NewLine);
       _locationInEntityCoordinates.reflection(sb);
           sb.Append("byte\t _detonationResult\t " + _detonationResult.ToString() + System.Environment.NewLine);
           sb.Append("byte\t _articulationParameters\t " + _articulationParameters.Count.ToString() + System.Environment.NewLine);
           sb.Append("short\t _pad\t " + _pad.ToString() + System.Environment.NewLine);

       for(int idx = 0; idx < _articulationParameters.Count; idx++)
       {
           sb.Append("ArticulationParameter\t " + _articulationParameters[idx] + System.Environment.NewLine);
            ArticulationParameter aArticulationParameter = (ArticulationParameter)_articulationParameters[idx];
            aArticulationParameter.reflection(sb);
       } // end of list marshalling

    } // end try 
    catch(Exception e)
    { 
      Trace.WriteLine(e);
      Trace.Flush();
}
    } // end of marshal method

 /**
  * The equals method doesn't always work--mostly it works only on on classes that consist only of primitives. Be careful.
  */
 public bool equals(DetonationPdu rhs)
 {
     bool ivarsEqual = true;

    if(rhs.GetType() != this.GetType())
        return false;

     if( ! (_munitionID.Equals( rhs._munitionID) )) ivarsEqual = false;
     if( ! (_eventID.Equals( rhs._eventID) )) ivarsEqual = false;
     if( ! (_velocity.Equals( rhs._velocity) )) ivarsEqual = false;
     if( ! (_locationInWorldCoordinates.Equals( rhs._locationInWorldCoordinates) )) ivarsEqual = false;
     if( ! (_burstDescriptor.Equals( rhs._burstDescriptor) )) ivarsEqual = false;
     if( ! (_locationInEntityCoordinates.Equals( rhs._locationInEntityCoordinates) )) ivarsEqual = false;
     if( ! (_detonationResult == rhs._detonationResult)) ivarsEqual = false;
     if( ! (_numberOfArticulationParameters == rhs._numberOfArticulationParameters)) ivarsEqual = false;
     if( ! (_pad == rhs._pad)) ivarsEqual = false;

     for(int idx = 0; idx < _articulationParameters.Count; idx++)
     {
        ArticulationParameter x = (ArticulationParameter)_articulationParameters[idx];
        if( ! ( _articulationParameters[idx].Equals(rhs._articulationParameters[idx]))) ivarsEqual = false;
     }


    return ivarsEqual;
 }
} // end of class
} // end of namespace
